File: trunk/+tom/OptimizerDefault.cpp
#include "Optimizer.h"

namespace tom
{
  /** This default optimizer does nothing and provides no solutions. */
  class OptimizerDefault : virtual public Optimizer
  {
  public:
    /** Constructor and parent class initializer of the same form. */
    OptimizerDefault(void) :
      Optimizer()
    {
      return;
    }

    uint32_t numInitialConditions(void) const
    {
      return (0);
    }

    void defineProblem(std::vector<DynamicModel::Handle>& dynamicModel, std::vector<Measure::Handle>& measure,
      bool randomize)
    {
      return;
    }

    void refreshProblem(void)
    {
      return;
    }

    uint32_t numSolutions(void)
    {
      return (0);
    }

    Trajectory* getSolution(const uint32_t k)
    {
      static Trajectory* x = NULL;
      throw("The default optimizer provides no solutions.");
      return (x);
    }

    double getCost(const uint32_t k)
    {
      throw("The default optimizer provides no solutions.");
      return (0.0);
    }

    void step(void)
    {
      return;
    }

  private:
    static std::string componentDescription(void)
    {
      return ("This default optimizer does nothing and provides no solutions.");
    }

    static Optimizer* componentFactory(void)
    {
      return (new OptimizerDefault());
    }

  protected:
    static void initialize(std::string name)
    {
      connect(name, componentDescription, componentFactory);
    }
    friend class OptimizerDefaultInitializer;
  };

  /** This class initializes OptimizerDefault before the execution of main(). */
  class OptimizerDefaultInitializer
  {
  public:
    OptimizerDefaultInitializer(void)
    {
      OptimizerDefault::initialize("tom");
    }
  } _OptimizerDefaultInitializer;
}

File: trunk/+tom/OptimizerDefault.m
classdef OptimizerDefault < tom.Optimizer
  
  methods (Static = true, Access = public)
    function initialize(name)
      function text = componentDescription
        text = 'This default optimizer does nothing and provides no solutions.';
      end
      tom.Optimizer.connect(name, @componentDescription, @tom.OptimizerDefault);
    end
  end
 
  methods (Access = public, Static = true)
    function this = OptimizerDefault()
      this = this@tom.Optimizer();
    end
  end
  
  methods (Access = public)  
    function num = numInitialConditions(this)
      assert(isa(this, 'tom.Optimizer'));
      num = uint32(0);
    end
    
    function defineProblem(this, dynamicModel, measure, randomize)
      assert(isa(this, 'tom.Optimizer'));
      assert(isa(dynamicModel, 'tom.DynamicModel'));
      assert(isa(measure, 'cell'));
      assert(isa(randomize, 'logical'));
    end
    
    function refreshProblem(this)
      assert(isa(this, 'tom.Optimizer'));
    end
    
    function num = numSolutions(this)
      assert(isa(this, 'tom.Optimizer'));
      num = uint32(0);
    end
    
    function x = getSolution(this, k)
      assert(isa(this, 'tom.Optimizer'));
      assert(isa(k, 'uint32'));
      x = tom.DynamicModelDefault(tom.WorldTime(0) ,'');
      assert(isa(x, 'tom.Trajectory'));
      error('The default optimizer provides no solutions.');
    end

    function cost = getCost(this, k)
      assert(isa(this, 'tom.Optimizer'));
      assert(isa(k, 'uint32'));
      cost = 0.0;
      assert(isa(cost, 'double'));
      error('The default optimizer provides no solutions.');
    end
    
    function step(this)
      assert(isa(this, 'tom.Optimizer'));
    end
  end
  
end
